packages/actor-abstract-path/lib/ActorAbstractPath.ts

```diff
@@ -73,15 +73,52 @@ export abstract class ActorAbstractPath extends ActorQueryOperationTypedMediated
 
   // Based on definition in spec https://www.w3.org/TR/sparql11-query/
   // returns all nodes visited by infinitely repeating the given predicate, starting from x
-  public async ALPeval(x: Term, predicate: Algebra.PropertyPathSymbol, context: ActionContext):
-  Promise<AsyncIterator<Term>> {
+  public async ALPeval(subject: Term, object: Variable, objectVal: Term, predicate: Algebra.PropertyPathSymbol,
+    graph: Term, context: ActionContext):
+    Promise<AsyncIterator<Bindings>> {
+    if (graph.termType === 'Variable') {
+      const findGraphs = ActorAbstractPath.FACTORY.createPath(subject, predicate, object, graph);
+      const results = ActorQueryOperation.getSafeBindings(
+        await this.mediatorQueryOperation.mediate({ context, operation: findGraphs }),
+      );
+
+      const objectString = termToString(object);
+
+      return new MultiTransformIterator(
+        results.bindingsStream,
+        {
+          multiTransform: (bindings: Bindings) => {
+            const graphValue = bindings.get(termToString(graph));
+            return new TransformIterator<Bindings>(
+              async() => {
+                const it = new BufferedIterator<Term>();
+                await this.ALP(objectVal, predicate, graphValue, context, {}, it, { count: 0 });
+                return it.transform<Bindings>({
+                  transform(item, next, push) {
+                    push(Bindings({ [objectString]: item, [termToString(graph)]: graphValue }));
+                    next();
+                  },
+                });
+              }, { maxBufferSize: 128 },
+            );
+          },
+          autoStart: false,
+        },
+      );
+    }
+
     const it = new BufferedIterator<Term>();
-    await this.ALP(x, predicate, context, {}, it, { count: 0 });
+    await this.ALP(objectVal, predicate, graph, context, {}, it, { count: 0 });
 
-    return it;
+    return it.transform<Bindings>({
+      transform(item, next, push) {
+        push(Bindings({ [termToString(object)]: item }));
+        next();
+      },
+    });
```

Comment:
Would not add this (and the extra function parameters for this) to the ALP function and create an external function to do this part since this is changing what the ALP function does compared to the [spec](https://www.w3.org/TR/sparql11-query/#PropertyPathPatterns) (or rename the function I guess).

Reply:
Suggestion for the new name of the function?

Reply:
And then maybe rename ALPeval to thatNamEval.

Reply:
No opinion, it's just that if a function is called ALP/ALPeval it should (sort of) do what the corresponding function in the spec does. Giving it a readable name that sort of explains what it does might be better anyway.